### DISCUSSION TITLE
Point where nlp scaling is done

### DISCUSSION THREAD

**Discussion by bradbell** [coin-or/Ipopt#481] [Q&A] [answered]
It seems from the `nlp_scaling_method` documentation
https://coin-or.github.io/Ipopt/OPTIONS.html#OPT_nlp_scaling_method
that the scaling depends on the starting point.

Often this is not what one wants; e.g., when warm starting
https://list.coin-or.org/pipermail/ipopt/2016-May/004221.html

Is there some way to specify a scaling point that is different from the starting point ?

**Comment by svigerske** [comment, marked as answer]
Not in a clean way.
You may have notices that `TNLP::get_starting_point()` is called twice. The first one is the one triggered by the gradient or equilibrium scaling methods. So in principle you could provide the point where to scale first and the actual starting point in the next call.

A better implementation could be to add another callback to `TNLP` (which default to `get_starting_point()` if not overwritten) or to add a flag to `get_starting_point()` to indicate whether the point where to compute the scaling is request.

However, the motivation for the warm starting is actually that you want to ensure that the same scaling is used for the warmstarted solve (that's why Nai-Yuan disabled scaling in this post). As a user can already provide the scaling, maybe one rather wants a way to retrieve the scaling parameters that were used by Ipopt, so one can pass them in again. Just providing the same scaling point could result in different scaling parameters for the objective and constraints that were modified. At least it is possible to retrieve the scaling parameters for the Ipopt-internal formulation of the NLP, so with some `TNLPAdapter` functions, one can translate this back into the `TNLP` space (or more crude: get scaled and unscaled values via `TNLP::get_curr_iterate()` (new in Ipopt 3.14) and reconstruct scaling values from them).

I haven't checked whether calling `ReOptimizeTNLP()` would already be sufficient to ensure that the scaling object from the previous solve will be reused. There is also a (hidden) parameter [warm_start_same_structure](https://github.com/coin-or/Ipopt/blob/stable/3.13/src/Algorithm/IpOrigIpoptNLP.cpp#L70-L77) which, if enabled, prevents the scaling to be recomputed (https://github.com/coin-or/Ipopt/blob/stable/3.13/src/Algorithm/IpOrigIpoptNLP.cpp#L232-L290).

**Comment by bradbell** [comment]
Reading the documentation, I could not figure out how to do this for my Ipopt applications. I am currently  doing my own scaling (before ipopt sees the function) and using `none` for the ipopt scaling. The scaling I do uses function values and gradients at a scaling point, which is different from the staring point.

It just seemed to me that it would be useful for ipopt to have an option like this; i.e., this is a feature request.
